### DISCUSSION TITLE
Updating PackageReferences

### DISCUSSION THREAD

**Discussion by AlanHoiland** [cake-build/cake#4130] [Q&A]
Hello,
I thought this would be easy but I'm missing something. How do you update PackageReferences version numbers?
Visual Studio 2022
netstandard 2.0 projects
I have a Shared package that is used in other projects. I can update the Shared package version, build, pack, and publish. But then want to update the PackageReference in another project and have hit a brick wall. I've tried NuGetUpdate and of course it is looking for a packages.config file that doesn't exist. I've tried Cake.Incubator and it can get the PackageReferences but I don't see a way to write any changes back to the project. I feel like I'm missing something basic but all my google searches turn up empty.

Here's my cake script:
///////////////////////////////////////////////////////////////////////////////
// ARGUMENTS
///////////////////////////////////////////////////////////////////////////////
#tool Bumpy
#addin Cake.Bumpy
#addin nuget:?package=Cake.ExtendedNuGet&version=5.0.0
#addin nuget:?package=Cake.Incubator&version=8.0.0

var target = Argument("target", "Default");
var configuration = Argument("configuration", "Release");

///////////////////////////////////////////////////////////////////////////////
// SETUP / TEARDOWN
///////////////////////////////////////////////////////////////////////////////

Setup(ctx =>
{
	// Executed BEFORE the first task.
	Information("Running tasks...");
});

Teardown(ctx =>
{
	// Executed AFTER the last task.
	Information("Finished running tasks.");
});

///////////////////////////////////////////////////////////////////////////////
// TASKS
///////////////////////////////////////////////////////////////////////////////

var versionTask = Task("Version")
.Does(() => 
{
	BumpyNew();

	BumpyIncrement(3, new BumpySettings{
		Directory = "./Shared"
		});
});

var cleanTask = Task("Clean")
.IsDependentOn(versionTask)
.Does(() =>
{
	CleanDirectory("./Shared/bin/release");
	CleanDirectory("./Shared/bin/debug");
});

var restoreTask = Task("Restore")
.IsDependentOn(cleanTask)
.Does(() =>
{
	DotNetRestore("./");
});

var buildTask = Task("Build")
.IsDependentOn(restoreTask)
.Does(() =>
{
	DotNetBuild("./Shared", new DotNetBuildSettings{
		NoRestore = true
		});
});

var packageTask = Task("Package")
.IsDependentOn(buildTask)
.Does(() =>
{
	DotNetPack("./Shared", new DotNetPackSettings
									{
										NoRestore = true,
										NoBuild = true
									});
});

var publishTask = Task("Publish")
.IsDependentOn(packageTask)
.Does(() => {
	var settings = new DotNetNuGetPushSettings
	{
		Source = "mysourcepath",
		ApiKey = "ApiKey"
	};

    var pkgs = GetFiles("./Shared/bin/debug/" + "*.nupkg");
	Information(pkgs.Count());

    foreach(var pkg in pkgs) 
    {
		Information(pkg);

        if(!IsNuGetPublished(pkg, settings.Source)) 
        {
            Information($"Publishing \"{pkg}\".");
            DotNetNuGetPush(pkg.FullPath, settings);
        }
        else {
            Information($"Bypassing publishing \"{pkg}\" as it is already published.");
        }
            
    }
});

var updateSharedTask = Task("UpdateShared")
//.IsDependentOn(publishTask)
.Does(() =>
{
	var props = ParseProject("./Project1/Project1.csproj", "Debug");
	
	Information(props.PackageReferences.Count());

	foreach(var project in props.PackageReferences)
	{
		Information($"{project.Name} {project.Version}");
	}

	// This doesn't work.
//	NuGetUpdate(@"./Library.sln", new NuGetUpdateSettings
//	{
//		Id = new [] { "Shared" }
//	});
});

Task("Default")
.IsDependentOn(updateSharedTask)
.Does(() => {
	Information("Hello Cake!");
});

RunTarget(target);

**Comment by augustoproiete** [comment]
@AlanHoiland The recommended way to update package references via command-line is to use the `dotnet add package` command

[![image](https://user-images.githubusercontent.com/177608/230742483-7323c953-86cd-4028-9c0b-e59ae3f88aad.png)](https://www.nuget.org/packages/NaturalStringExtensions)

---

e.g. `MyProject.csproj` references [NaturalStringExtensions](https://www.nuget.org/packages/NaturalStringExtensions) v2.1.0 and you want to bump it to v3.0.0:

- `MyProject.csproj`

```
<Project Sdk="Microsoft.NET.Sdk">
  <ItemGroup>
    <PackageReference Include="NaturalStringExtensions" Version="2.1.0" />
  </ItemGroup>
</Project>
```

---

- `build.cake`

```cake
StartProcess("dotnet", "add package NaturalStringExtensions --version 3.0.0");
```
